src/python/py_parsers.cpp

```diff
@@ -0,0 +1,221 @@
+// Copyright (C) 2023-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include <filesystem>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <pybind11/stl_bind.h>
+#include <pybind11/functional.h>
+
+#include "openvino/genai/parsers.hpp"
+#include "py_utils.hpp"
+#include "openvino/genai/json_container.hpp"
+
+namespace py = pybind11;
+
+using ov::genai::IncrementalParserBase;
+using ov::genai::ParserBase;
+using ov::genai::ReasoningParser;
+using ov::genai::Phi4ReasoningParser;
+using ov::genai::DeepSeekR1ReasoningParser;
+using ov::genai::JsonContainer;
+using ov::genai::Llama32JsonToolParser;
+using ov::genai::Llama32PythonicToolParser;
+using ov::genai::Tokenizer;
+using ov::genai::StreamingStatus;
+
+namespace pyutils = ov::genai::pybind::utils;
+
+namespace {
+
+
+class ConstructableIncrementalParserBase: public IncrementalParserBase {
+public:
+    std::string parse(
+        JsonContainer& msg,
+        const std::string& previous_text, 
+        std::string& delta_text, 
+        const std::optional<std::vector<int64_t>>& previous_tokens = std::nullopt, 
+        const std::optional<std::vector<int64_t>>& delta_tokens = std::nullopt
+    ) override {
+        PYBIND11_OVERRIDE_PURE(
+            std::string,  // Return type
+            IncrementalParserBase,  // Parent class
+            parse,  // Name of function in C++ (must match Python name)
+            msg,
+            previous_text,
+            delta_text,
+            previous_tokens,
+            delta_tokens
+        );
+    }
+};
+
+class ConstructableParserBase: public ParserBase {
+public:
+    void parse(JsonContainer& text) override {
+        PYBIND11_OVERRIDE_PURE(
+            void,  // Return type
+            ParserBase,  // Parent class
+            parse,  // Name of function in C++ (must match Python name)
+            text  // Argument(s)
+        );
+    }
+};
+
+
+// wrapper to enhance calling parser from Python
+void call_parser(py::dict& msg, std::function<void(JsonContainer&)> func) {
+    auto msg_cpp = pyutils::py_object_to_json_container(msg);
+    func(msg_cpp);
+
+    // TODO: msg = pyutils::json_container_to_py_object(msg_cpp) does not work properly here,
+    py::object json_mod = py::module_::import("json");
+    
+    // since it create a new object instead of updating existing dict.
+    auto json_str = msg_cpp.to_json_string();
+    py::dict result = json_mod.attr("loads")(json_str);
+    
+    // update msg with result
+    msg.clear();
+    for (auto item : result) {
+        msg[item.first] = item.second;
+    }
```

Comment:
Can we use py::dict update() method?
```suggestion
    msg.update(result);
```